### DISCUSSION TITLE
Any way to use pre-computed cell embeddings as input for dyn.tl.moments?

### DISCUSSION THREAD

**Discussion by altairwei** [aristoteleo/dynamo-release#421] [Q&A] [answered]
In scvelo, I can use the `use_rep` argument of `scvelo.pp.moments` to compute first-/second-order moments based on the given cell embeddings. Many batch effect correcting tools are operated on dimensionality reduction space, such as `harmony` or `fastMNN`. It's important for multi-sample scRNA-Seq data to carry out downstream analysis based on batch-corrected cell embeddings.

However, preprocessing steps in dynamo are wrapped within one function `recipe_monocle`, there is no way to break it down and plug my batch-corrected cell embeddings in it. Is there any workaround fot the problem?

**Comment by Xiaojieqiu** [comment, marked as answer]
hey @altairwei , you can use` dyn.tl.moments(adata, X_data=adata.obsm['X_harmony']) `or others other embedding in `.obsm` to calculate first/second and mix-order moments with dynamo. 

Dynamo's preprocessing module is actually very flexible and we have integrated Pearson residual and harmony in our model. Unfortunately, I am busy with job searches these days, but you can check this notebook to get it started: https://github.com/aristoteleo/dynamo-tutorials/blob/master/tutorial_preprocess.ipynb

We will have a tutorial on preprocessing in our readthedocs page sometime next month.

**Comment by altairwei** [comment]
Hi @Xiaojieqiu, [tutorial_preprocess.ipynb](https://github.com/aristoteleo/dynamo-tutorials/blob/master/tutorial_preprocess.ipynb) is great, and it made me realize that dynamo's preprocessing is indeed customizable. I am interested in dynamo's `negbin` mdoe for conventional scRNA-seq, but I just want to normalize spliced&unspliced data then calculate moments based on harmony cell embeddings (the customized preprocessor part is not yet written in the tutorial). However, the source codes of `dyn.tl.dynamics` are complex, and I can't find a way to separate the logic of moments calculation from it. That makes it seem unlikely to calculate moments based on given cell embeddings. Could you give me more hints?

**Reply by Xiaojieqiu** [reply]
please see my above code on running `dyn.tl.moments`. Once moments calculated before dynamics, it won't be recomputed

**Reply by altairwei** [reply]
Gosh, I totally didn't realize that! Thank you so much!
